kore/src/Debug.hs

```diff
@@ -0,0 +1,581 @@
+    => a
+    -> a
+    -> Maybe (Int -> Doc ann)
+diffPrecGeneric a b =
+    diffPrecSOP (SOP.datatypeInfo (Proxy @a)) (a, SOP.from a) (b, SOP.from b)
+
+diffPrecSOP
+    :: forall a xss ann
+    .  (Debug a, All2 Diff xss)
+    => DatatypeInfo xss
+    -> (a, SOP I xss)
+    -> (a, SOP I xss)
+    -> Maybe (Int -> Doc ann)
+diffPrecSOP datatypeInfo (a, SOP aNS) (b, SOP bNS) =
+    diffNS constrs aNS bNS
+  where
+    constrs :: NP ConstructorInfo xss
+    constrs =
+        case datatypeInfo of
+            SOP.ADT     _ _ cs -> cs
+            SOP.Newtype _ _ c  -> c :* Nil
+
+    diffNS
+        :: forall xss'
+        .  All2 Diff xss'
+        => NP ConstructorInfo xss'
+        -> NS (NP I) xss'
+        -> NS (NP I) xss'
+        -> Maybe (Int -> Doc ann)
+    diffNS (c :* _) (Z aNP) (Z bNP) = diffNP c aNP bNP
+    diffNS (_ :* cs) (S aNS') (S bNS') = diffNS cs aNS' bNS'
+    diffNS _ _ _ =
+        Just $ \precOut ->
+            Pretty.sep
+                [ "{- was:"
+                , debugPrec a precOut
+                , "-}"
+                , debugPrec b precOut
+                ]
+
+    diffNP
+        :: forall xs
+        .  All Diff xs
+        => ConstructorInfo xs
+        -> NP I xs
+        -> NP I xs
+        -> Maybe (Int -> Doc ann)
+    diffNP c aNP bNP
+      | anyNP (isJust . SOP.unK) cNP =
+        Just $ SOP.unK $ debugConstr c (SOP.hmap (SOP.mapKK maybeHole) cNP)
+      | otherwise =
+        Nothing
+      where
+        cNP = diffNP' aNP bNP
+        maybeHole = fromMaybe (const "_")
+
+    anyNP :: forall f xs. (forall x. f x -> Bool) -> NP f xs -> Bool
+    anyNP query (fx :* fxs) = query fx || anyNP query fxs
+    anyNP _ Nil = False
+
+    diffNP'
+        :: forall xs
+        .  All Diff xs
+        => NP I xs
+        -> NP I xs
+        -> NP (K (Maybe (Int -> Doc ann))) xs
+    diffNP' = SOP.hczipWith (Proxy @Diff) (SOP.mapIIK diffPrec)
+
+instance Diff Bool where
+    diffPrec = diffPrecEq
+
+instance (Debug a, Diff a) => Diff [a]
+
+instance {-# OVERLAPS #-} Diff String where
+    diffPrec = diffPrecEq
+
+instance Diff Text where
+    diffPrec = diffPrecEq
+
+instance Diff Void where
+    diffPrec = diffPrecEq
+
+instance Diff () where
+    diffPrec = diffPrecEq
+
+instance Diff Natural where
+    diffPrec = diffPrecEq
+
+instance Diff Integer where
+    diffPrec = diffPrecEq
+
+instance Diff Int where
+    diffPrec = diffPrecEq
+
+instance Diff Char where
+    diffPrec = diffPrecEq
+
+instance (Debug a, Diff a) => Diff (Const a b)
+
+instance (Debug a, Diff a) => Diff (Maybe a)
+
+instance (Debug a, Diff a) => Diff (Sup a)
+
+instance (Debug a, Diff a) => Diff (Identity a)
+
+instance
+    (Debug a, Debug (f b), Diff a, Diff (f b)) => Diff (CofreeF f a b) where
+    diffPrec x y =
+        -- Cannot have orphan instances of Generic and HasDatatypeInfo.
+        -- Use a fake instance instead.
+        diffPrecSOP datatypeInfoCofreeF (x, fromCofreeF x) (y, fromCofreeF y)
```

Comment:
How hard would it be to diff the annotation last?

Reply:
You mean, print the difference in the annotations only if the patterns are the same?

Reply:
Yes, that's what I meant. I wasn't sure exactly how this works until I saw the additional tests. I'm not asking you to include it in this PR, but it may turn out to be useful.